contracts/masset/Masset.sol

```diff
@@ -573,46 +644,138 @@ contract Masset is
+                    vaultBalance: args.bAssets[i].vaultBalance
+                })
+            );
+            // 2.1. Log the net amounts (output - fee)
+            netAmounts[i] = amt.net;
+            // 2.2. Accumulate scaled fees
+            fees = fees.add(amt.scaledFee);
+        }
+        // 2.3. Log the collected fees to the surplus
+        surplus = args.cache.surplus.add(fees);
+
+        // 3.0. Reduce the vaultBalances by the **net** amount
+        basketManager.decreaseVaultBalances(args.indices, args.integrators, netAmounts);
+    }
+
+    struct WithdrawArgs {
+        uint256 quantity;
+        address bAsset;
+        address integrator;
+        uint256 feeRate;
+        bool hasTxFee;
+        address recipient;
+        uint256 ratio;
+        uint256 maxCache;
+        uint256 vaultBalance;
+    }
+
+    /**
+     * @dev Withdraws a given asset from its platformIntegration. If there is sufficient liquidity
+     * in the cache, then withdraw from there, otherwise withdraw from the lending market and reset the
+     * cache to the mid level.
+     * @param args     All args needed for a full withdrawal
+     * @return amount  Struct containing the desired output, output-fee, and the scaled fee
+     */
+    function _withdrawTokens(WithdrawArgs memory args) internal returns (Amount memory amount) {
+        console.log("_withdrawTokens: q = args.quantity");
+        if(args.quantity > 0){
+
+            // 1. Deduct the redemption fee, if any, and log quantities
+            amount = _deductSwapFee(args.bAsset, args.quantity, args.feeRate, args.ratio);
+
+            // 2. If txFee then short circuit - there is no cache
+            if(args.hasTxFee){
+                console.log("_withdrawTokens: hasTxFee");
+                IPlatformIntegration(args.integrator).withdraw(args.recipient, args.bAsset, amount.net, amount.net, true);
+            }
+            // 3. Else, withdraw from either cache or main vault
+            else {
+                uint256 cacheBal = IERC20(args.bAsset).balanceOf(args.integrator);
+                // 3.1 - If balance b in cache, simply withdraw
+                if(cacheBal > amount.net) {
+                    console.log("_withdrawTokens: cacheBal > net - '%s' > '%s'", cacheBal, amount.net);
+                    IPlatformIntegration(args.integrator).withdrawRaw(args.recipient, args.bAsset, amount.net);
+                }
+                // 3.2 - Else reset the cache to X, or as far as possible
+                //       - Withdraw X+b from platform
+                //       - Send b to user
+                else {
+                    console.log("_withdrawTokens: cacheBal < net - '%s' < '%s'", cacheBal, amount.net);
+                    uint256 relativeMidCache = args.maxCache.divRatioPrecisely(args.ratio).div(2);
+                    uint256 totalWithdrawal = StableMath.min(relativeMidCache.sub(cacheBal).add(amount.net), args.vaultBalance.sub(cacheBal));
```

Comment:
Is it possible for `relativeMidCache` to be less than `cacheBal`? Seems like it might be possible if the cache size was decreased earlier. It would be good to make sure that none of the safemath operations revert (maybe by changing the order).

Another question: What's the reason for considering `args.vaultBalance.sub(cacheBal)`? It seems like `vaultBalance` might be less than `cacheBal` if someone transferred tokens to the cache.

Reply:
Fixed the LHS by swapping the add and subtract around, since amount.net is > cacheBal.

RHS - vaultBalance.sub(cacheBal) must be considered because the total withdrawal can at maximum be everything that is in the lending market (this is the vaultbalance minus what is in the cache and does not include any interest that has been generated).

If the cacheBal is > vaultBalance, we simply cannot reach this point, and so this can be left as it is. Why? Because if the previous statement 'if(cacheBal >= amount.net)'. If amount.net was greater than the cache balance, and the cache was greater than the vault balance, then the amount.net is greater than the vault balance... which would be picked up in the Forge Validator

Reply:
Thanks!

About the LHS: Actually, then why not even change the precedence from (relativeMidCache + amount.net) - cacheBal to relativeMidCache + (amount.net - cacheBal)? That seems even slightly more easy to reason about.

About the RHS: I see. Yeah, I think this makes sense. I wonder if you can use a similar argument to show that LHS <= RHS, thus making the min-operation unnecessary.

Reply:
LHS - agree on the rationale, updated.

RHS - the min operation **is** necessary.. LHS > RHS in the majority of cases.